File: ios/Classes/LauncherShortcutsPlugin.swift
import Flutter
import UIKit

/// A protocol defining the contract for an object that can provide and manage
/// an array of `UIApplicationShortcutItem` instances.
///
/// This abstraction is primarily used to facilitate testing by allowing
/// mock implementations of shortcut item sources.
protocol ShortcutItemProviding: AnyObject {
    /// The application\'s dynamic shortcut items.
    ///
    /// Setting this property should update the application\'s displayed
    /// shortcuts. Getting it should return the currently set shortcuts.
    var shortcutItems: [UIApplicationShortcutItem]? { get set }
}

/// Extends `UIApplication` to conform to the `ShortcutItemProviding` protocol.
///
/// This allows `UIApplication.shared` to be used as the default provider
/// for application shortcut items, interacting directly with the iOS system.
extension UIApplication: ShortcutItemProviding {}

/// The main plugin class for the `launcher_shortcuts` Flutter plugin on iOS.
///
/// This class is responsible for:
/// - Establishing communication with the Flutter side of the plugin.
/// - Managing dynamic application shortcut items.
/// - Responding to application lifecycle events related to shortcut
///   activation.
public final class LauncherShortcutsPlugin: NSObject, FlutterPlugin, ShortcutsApi {
    /// The API used for sending messages and invoking methods on the Flutter
    /// side.
    /// This is used to notify Flutter about shortcut activations.
    private let flutterApi: IOSShortcutsApiProtocol

    /// The provider for `UIApplicationShortcutItem` instances.
    ///
    /// This is abstracted via the `ShortcutItemProviding` protocol to allow
    /// for dependency injection, primarily for testing purposes. It defaults
    /// to `UIApplication.shared`.
    private let shortcutProvider: ShortcutItemProviding

    /// Stores the `type` string of a shortcut item if the application was
    /// launched from a terminated state by that shortcut.
    ///
    /// This value is captured in `application(_:didFinishLaunchingWithOptions:)`
    /// and processed in `applicationDidBecomeActive(_:)`. It is cleared after
    /// being processed.
    private var launchingShortcutType: String?

    /// A flag indicating whether the Flutter side has signaled that it is
    /// initialized and ready to receive shortcut action events.
    /// Defaults to `false`.
    private var isFlutterReady = false

    /// Stores a shortcut action `type` string if it was received (e.g., via
    /// `performActionFor`) before Flutter was ready to handle it.
    ///
    /// This action is processed once `isFlutterReady` becomes `true`.
    private var pendingAction: String?

    /// A static reference to the plugin instance.
    ///
    /// This allows other parts of the native application (e.g., AppDelegate)
    /// to access the plugin instance if necessary, though this is generally
    /// not required for typical plugin operation.
    private static var instance: LauncherShortcutsPlugin?

    /// Registers this plugin with the Flutter engine.
    ///
    /// This static method is called by Flutter when the plugin is first
    /// registered. It sets up the method channel, application delegate,
    /// and initializes the plugin instance.
    ///
    /// - Parameter registrar: The `FlutterPluginRegistrar` provided by the
    ///   Flutter engine, used for accessing the binary messenger, registering
    ///   delegates, etc.
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        // Initialize the Flutter API for communication back to Dart.
        let flutterApi = IOSShortcutsApi(binaryMessenger: messenger)
        // Create the plugin instance with the Flutter API.
        let instance = LauncherShortcutsPlugin(flutterApi: flutterApi)
        // Set up the Pigeon-generated API for calls from Dart to native.
        ShortcutsApiSetup.setUp(binaryMessenger: messenger, api: instance)
        // Register the plugin as an application delegate to receive lifecycle
        // events.
        registrar.addApplicationDelegate(instance)
        // Store the static instance.
        self.instance = instance
    }

    /// Initializes a new plugin instance.
    ///
    /// - Parameters:
    ///   - flutterApi: An object conforming to `IOSShortcutsApiProtocol`, used
    ///     for sending messages to the Flutter side.
    ///   - shortcutProvider: An object conforming to `ShortcutItemProviding`,
    ///     used to manage the application\'s shortcut items. Defaults to
    ///     `UIApplication.shared`.
    init(
        flutterApi: IOSShortcutsApiProtocol,
        shortcutProvider: ShortcutItemProviding = UIApplication.shared
    ) {
        self.flutterApi = flutterApi
        self.shortcutProvider = shortcutProvider
        super.init()
    }

    // MARK: - ShortcutsApi Implementation

    /// Sets the dynamic shortcut items for the application.
    ///
    /// This method is called from the Flutter side to update the app\'s
    /// dynamic shortcuts. It maps the provided `ShortcutItem` DTOs to
    /// `UIApplicationShortcutItem` instances.
    ///
    /// - Parameters:
    ///   - itemsList: An array of `ShortcutItem` data transfer objects,
    ///     each representing a shortcut to be set.
    ///   - completion: A closure that is called with the result of the
    ///     operation. It receives `.success(())` if the items are set
    ///     (or an empty list is processed), or an error if one occurs
    ///     (though this implementation currently does not produce errors).
    func setShortcutItems(
        itemsList: [ShortcutItem],
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        shortcutProvider.shortcutItems = itemsList.compactMap { item in
            // Map the platform-agnostic icon name to a
            // UIApplicationShortcutIcon if provided.
            let icon = item.iosConfig?.icon.map {
                UIApplicationShortcutIcon(templateImageName: $0)
            }
            return UIApplicationShortcutItem(
                type: item.type,
                localizedTitle: item.localizedTitle,
                localizedSubtitle: item.iosConfig?.localizedSubtitle,
                icon: icon,
                userInfo: nil  // UserInfo is not currently used.
            )
        }
        completion(.success(()))
    }

    /// Clears all dynamic shortcut items for the application.
    ///
    /// This method is called from the Flutter side. It conforms to a
    /// potentially throwing API contract from Pigeon, but this specific
    /// implementation does not currently throw any errors.
    func clearShortcutItems() throws {
        shortcutProvider.shortcutItems = []
    }

    /// Notifies the plugin that the Flutter side is initialized and ready to
    /// handle shortcut actions.
    ///
    /// If a `pendingAction` (a shortcut action received before Flutter was
    /// ready) exists, it will be processed immediately.
    ///
    /// - Parameter completion: A closure that is called with the result of the
    ///   operation. It receives `.success(())` upon completion.
    func setFlutterReady(completion: @escaping (Result<Void, Error>) -> Void) {
        isFlutterReady = true
        // If there was an action waiting for Flutter to be ready, handle it now.
        if let action = pendingAction {
            handleShortcut(action)
            pendingAction = nil  // Clear the pending action after handling.
        }
        completion(.success(()))
    }

    // MARK: - UIApplicationDelegate Implementation

    /// Handles the scenario where the app is launched from a terminated state
    /// via a shortcut item.
    ///
    /// This method checks the `launchOptions` for a shortcut item. If found,
    /// it stores the shortcut\'s `type` and returns `false` to prevent
    /// `application(_:performActionFor:completionHandler:)` from being
    /// called redundantly.
    ///
    /// - Parameters:
    ///   - application: The singleton `UIApplication` object.
    ///   - launchOptions: A dictionary indicating the reason the app was
    ///     launched. This may contain the `UIApplication.LaunchOptionsKey
    ///     .shortcutItem` key.
    /// - Returns: `false` if the app was launched by a shortcut item,
    ///   indicating that the shortcut launch has been acknowledged. `true`
    ///   otherwise.
    public func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions:
            [UIApplication.LaunchOptionsKey: Any]? = nil
    ) -> Bool {
        if let shortcutItem = launchOptions?[.shortcutItem]
            as? UIApplicationShortcutItem
        {
            // Store the type of the shortcut that launched the app.
            // This will be processed in `applicationDidBecomeActive`.
            launchingShortcutType = shortcutItem.type
            // Return false to indicate we\'ve handled the shortcut launch.
            // This prevents `performActionFor` from being called for this
            // initial launch shortcut.
            return false
        }
        return true
    }

    /// Handles shortcut item processing when the app becomes active.
    ///
    /// This method is called after `didFinishLaunchingWithOptions` (for cold
    /// starts) or when the app returns from the background. It processes any
    /// `launchingShortcutType` captured at launch or any `pendingAction`
    /// that was queued while Flutter was not ready.
    ///
    /// - Parameter application: The singleton `UIApplication` object.
    public func applicationDidBecomeActive(_ application: UIApplication) {
        // Process shortcut from a cold start, if any.
        if let type = launchingShortcutType {
            if isFlutterReady {
                handleShortcut(type)
            } else {
                // If Flutter is not ready yet, store this as a pending action.
                pendingAction = type
            }
            // Clear the launching shortcut type as it\'s been acknowledged.
            launchingShortcutType = nil
        }

        // Process any other pending action if Flutter is now ready.
        // This can happen if a shortcut was activated while the app was in the
        // background and Flutter wasn\'t ready, then became active.
        if let action = pendingAction, isFlutterReady {
            handleShortcut(action)
            pendingAction = nil
        }
    }

    /// Handles shortcut item selection when the app is already running (either
    /// in the foreground or background).
    ///
    /// If Flutter is ready, the shortcut action is handled immediately.
    /// Otherwise, it is stored as a `pendingAction`.
    ///
    /// - Parameters:
    ///   - application: The singleton `UIApplication` object.
    ///   - shortcutItem: The `UIApplicationShortcutItem` that was selected by
    ///     the user.
    ///   - completionHandler: A block to call with a Boolean value. `true`
    ///     indicates the shortcut action was handled (or will be handled).
    /// - Returns: Always returns `true` as the action is either handled
    ///   immediately or queued.
    public func application(
        _ application: UIApplication,
        performActionFor shortcutItem: UIApplicationShortcutItem,
        completionHandler: @escaping (Bool) -> Void
    ) -> Bool {
        if isFlutterReady {
            handleShortcut(shortcutItem.type)
        } else {
            // If Flutter is not ready, store the action to be handled later.
            pendingAction = shortcutItem.type
        }
        // Indicate that the shortcut action has been (or will be) handled.
        completionHandler(true)
        return true
    }

    // MARK: - Private Helper Methods

    /// Sends the shortcut action `type` to the Flutter side via the
    /// `flutterApi`.
    ///
    /// This method dispatches the call to the main thread to ensure UI-related
    /// Flutter calls are safe.
    ///
    /// - Parameter type: The `type` string of the shortcut item that was
    ///   activated.
    private func handleShortcut(_ type: String) {
        DispatchQueue.main.async {
            self.flutterApi.launchAction(action: type) { result in
                // The result of the Flutter call can be logged or handled
                // here if needed in the future (e.g., for error reporting).
                if case .failure(let error) = result {
                    // Example: Log error if Flutter side reports an issue.
                    NSLog("Error from Flutter launchAction: \(error)")
                }
            }
        }
    }

    /// Provides static access to the plugin instance.
    ///
    /// This can be used by other parts of the native iOS application (e.g.,
    /// `AppDelegate`) to interact with the plugin if necessary, although such
    /// direct interaction is generally discouraged in favor of Flutter-side
    /// control and standard plugin lifecycle management.
    ///
    /// - Returns: The singleton instance of `LauncherShortcutsPlugin`, or `nil`
    ///   if the plugin has not been registered or has been deallocated.
    public static func getInstance() -> LauncherShortcutsPlugin? {
        return instance
    }
}

File: android/src/main/kotlin/com/pco/camilo/launcher_shortcuts/LauncherShortcutsPlugin.kt
package com.pco.camilo.launcher_shortcuts

import android.app.Activity
import android.content.Intent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger

/**
 * [FlutterPlugin] and [ActivityAware] implementation for the AppShortcuts plugin.
 *
 * This plugin handles communication between Flutter and the native Android
 * platform for managing app shortcuts. It correctly manages lifecycle events
 * to ensure that communication channels and handlers are set up and torn down
 * at appropriate times.
 */
class LauncherShortcutsPlugin : FlutterPlugin, ActivityAware {

  /** Reference to the current Android [Activity]. Null if not attached. */
  private var activity: Activity? = null

  /**
   * [BinaryMessenger] for communication between Flutter and native platform.
   * Initialized in [onAttachedToEngine] and cleared in
   * [onDetachedFromEngine].
   */
  private var flutterMessenger: BinaryMessenger? = null

  /**
   * Instance of [LauncherShortcutsHandler] for handling Android-specific
   * shortcut logic. Initialized in [setupApis] when an [Activity] is
   * available and cleared in [onDetachedFromEngine].
   */
  private var androidApi: LauncherShortcutsHandler? = null

  /**
   * Instance of [AndroidShortcutsFlutterApi] for sending messages from native
   * to Flutter. Initialized in [onAttachedToEngine] and cleared in
   * [onDetachedFromEngine].
   */
  private var flutterApi: AndroidShortcutsFlutterApi? = null

  /**
   * Called when the plugin is attached to the Flutter engine.
   *
   * Initializes [flutterMessenger] and [flutterApi]. It then attempts to
   * set up the method channel APIs via [setupApis], which might fully
   * complete if an [Activity] is already available (e.g., if
   * [onAttachedToActivity] was called before this).
   *
   * @param binding The binding for Flutter plugin, providing access to
   *                engine components like the [BinaryMessenger].
   */
  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    // Initialize the messenger for communication with Flutter.
    flutterMessenger = binding.binaryMessenger

    // Initialize the Flutter API (native to Flutter calls) with the messenger.
    flutterApi = AndroidShortcutsFlutterApi(binding.binaryMessenger)

    // Attempt to setup the Android and Flutter APIs. This might not fully
    // initialize androidApi if the activity is not yet available.
    setupApis()
  }

  /**
   * Sets up the Android (Flutter to native) and Flutter (native to Flutter)
   * APIs for communication.
   *
   * This method initializes [androidApi] using the current [activity] and
   * sets up the method channel communication if both [activity] and
   * [flutterMessenger] are available. It ensures that [flutterApi] is
   * passed to [androidApi] for callbacks.
   */
  private fun setupApis() {
    // Ensure both activity and flutterMessenger are available before proceeding.
    activity?.let { act ->
      flutterMessenger?.let { messenger ->
        // Initialize the Android API handler with the current activity.
        androidApi = LauncherShortcutsHandler(act)

        // Link the Flutter API to the Android API.
        // flutterApi is guaranteed non-null here if flutterMessenger is
        // non-null, as it's initialized directly after flutterMessenger
        // in onAttachedToEngine.
        androidApi?.setFlutterApi(flutterApi!!)

        // Set up the method channel for Flutter to call Android methods.
        AndroidShortcutsApi.setUp(messenger, androidApi)
      }
    }
  }

  /**
   * Called when the plugin is detached from the Flutter engine.
   *
   * Clears all references to [flutterMessenger], [androidApi], and
   * [flutterApi] to ensure proper cleanup, break circular references,
   * and prevent memory leaks.
   *
   * @param binding The binding for Flutter plugin.
   */
  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    // Clear references to avoid memory leaks and ensure a clean state
    // if the plugin is re-attached later.
    flutterMessenger = null
    androidApi = null // This also helps break any potential cycles with activity.
    flutterApi = null
  }

  // ActivityAware Lifecycle Methods

  /**
   * Called when the plugin is attached to an [Activity].
   *
   * Initializes the local [activity] reference. It then registers a listener
   * for new intents (e.g., when the app is already running and a shortcut is
   * tapped). Crucially, it calls [setupApis] to establish communication
   * channels and then calls `androidApi?.handleIntent` to process the
   * intent that started the activity (important for cold starts via shortcuts).
   *
   * @param binding The binding for activity, providing access to the
   *                [Activity] and lifecycle callbacks.
   */
  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    // Initialize the activity reference.
    activity = binding.activity

    // Register listener for new intents. This handles cases where the app is
    // already running and receives a new intent (e.g., from a shortcut).
    binding.addOnNewIntentListener { intent ->
      // Ensure the activity is brought to the foreground if it exists
      intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or
              Intent.FLAG_ACTIVITY_CLEAR_TOP or
              Intent.FLAG_ACTIVITY_SINGLE_TOP

      // For launchMode="singleInstance" or "singleTask", the Android system
      // typically handles intent flags like FLAG_ACTIVITY_CLEAR_TOP.
      // This ensures the intent is passed to our handler.
      androidApi?.handleIntent(intent)

      // Indicate that the intent has been handled.
      true
    }

    // Setup the Android and Flutter APIs, now that an activity is available.
    setupApis()
    // Handle the initial intent that might have started this activity.
    // This is crucial for cold starts initiated by an app shortcut.
    androidApi?.handleIntent(binding.activity.intent)
  }

  /**
   * Called when the activity is detached from the plugin due to a
   * configuration change (e.g., screen rotation, theme change).
   *
   * Clears the local [activity] reference. The plugin will typically be
   * reattached to a new activity instance shortly via
   * [onReattachedToActivityForConfigChanges].
   */
  override fun onDetachedFromActivityForConfigChanges() {
    // Clear the activity reference as it's about to be destroyed and
    // replaced due to a configuration change.
    activity = null
  }

  /**
   * Called when the plugin is reattached to an [Activity] after a
   * configuration change.
   *
   * Re-initializes the local [activity] reference with the new activity
   * instance, re-registers the listener for new intents, and calls
   * [setupApis] to re-establish communication channels with the new
   * activity context.
   *
   * @param binding The binding for the new activity instance.
   */
  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    // Re-initialize the activity reference with the new instance.
    activity = binding.activity

    // Re-register listener for new intents with the new activity instance.
    binding.addOnNewIntentListener { intent ->
      // Similar to onAttachedToActivity, this ensures new intents are
      // processed by our handler after a configuration change.
      androidApi?.handleIntent(intent)
      true // Indicate that the intent has been handled.
    }

    // Re-setup the Android and Flutter APIs with the new activity.
    setupApis()
    // Generally, the activity's intent (binding.activity.intent) doesn't
    // need to be re-handled here as it's the same intent that caused the
    // original activity creation (or a subsequent onNewIntent), which should
    // have already been processed or will be if it's a new one via listener.
  }

  /**
   * Called when the [Activity] is detached from the plugin, and this is
   * not due to a configuration change (e.g., activity is being finished).
   *
   * Clears the local [activity] reference. If [androidApi] holds a direct
   * reference to the activity, it might become stale. However, [setupApis]
   * ensures [androidApi] is re-created with a valid activity if one
   * becomes available later. [androidApi] is also cleared in
   * [onDetachedFromEngine].
   */
  override fun onDetachedFromActivity() {
    // Clear the activity reference as the activity is being destroyed or the
    // plugin is no longer associated with it for other reasons.
    activity = null
  }
}
